Botnet linking with TLS (STARTTLS)

From the leaf's point of view:

<Domino> .link MyHub
<Leaf> [17:34:19] #Domino# link MyHub
<Leaf> [17:34:19] Linking to MyHub at <Hub's IP>:9348 ...
<Leaf> [17:34:19] Got STARTTLS from MyHub. Replying...
<Leaf> [17:34:19] Received challenge from MyHub... sending response ...
... and nothing more happens.

On the target (MyHub)
<MyHub> [02:38:00] Telnet connection: <Leaf's IP>/39848
<MyHub> [02:38:00] Timeout/EOF ident connection
<MyHub> [02:38:00] Sent STARTTLS to Leaf...
<MyHub> [02:38:00] Challenging Leaf...
<MyHub> [02:35:02] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:35:02] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:35:02] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:35:02] TLS: failed in: SSLv2/v3 read client hello A.
... this repeats until the leaf gives-up (timeout), APPROXIMATELY 40 TIMES !!
<MyHub> [02:39:01] Lost dcc connection to Leaf (telnet@<Leaf's IP>/39848)

Once this begins, I'm helpless to STOP it!

.DCCSTAT on the leaf provides:

<Leaf> 12 <MyHub's IP> + 9348 MyHub bot* waited 39s

<Domino> .tcl killdcc 12
<Leaf> Tcl error: invalid idx

Ummmm.... That's useless (and a lie)

Let's try on the MyHub!
<Domino> .dccstat
<Snip>
<MyHub> 25 <Leaf's IP> +38008 Leaf pass waited 25s
<MyHub> [02:41:26] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:41:27] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:41:28] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:41:29] TLS: failed in: SSLv2/v3 read client hello A.
<Domino> .tcl killdcc 25
<MyHub> [02:41:30] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> Tcl error: invalid idx
<MyHub> [02:41:31] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:41:32] TLS: failed in: SSLv2/v3 read client hello A.
<MyHub> [02:41:33] TLS: failed in: SSLv2/v3 read client hello A.

So, I can't registered DCC indeces, and must "wait it out" for a timeout.

Both bots have successfully connected to other bots (TLS via STARTTLS).

... if you care.

The following compiler warning may also be related and should be fixed:

```
gcc -g -O2 -pipe -Wall -I.. -I..  -DHAVE_CONFIG_H -I/usr/include -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -c dcc.c
dcc.c: In function ‘timeout_dupwait’:
dcc.c:1477:35: warning: ‘%s’ directive output may be truncated writing up to 323 bytes into a region of size between 67 and 99 [-Wformat-truncation=]
     egg_snprintf(x, sizeof x, "%s!%s", dcc[idx].nick, dcc[idx].host);
                                   ^~
In file included from compat/compat.h:27,
                 from main.h:102,
                 from dcc.c:26:
compat/snprintf.h:58:24: note: ‘snprintf’ output between 2 and 357 bytes into a destination of size 100
 #  define egg_snprintf snprintf
dcc.c:1477:5: note: in expansion of macro ‘egg_snprintf’
     egg_snprintf(x, sizeof x, "%s!%s", dcc[idx].nick, dcc[idx].host);
     ^~~~~~~~~~~~

```

I would prefer to first cleanup compat code, because snprintf is posix 2001 and then, fix the remaining compiler warning.

--- a/src/dcc.c
+++ b/src/dcc.c
@@ -1782,6 +1782,11 @@ static void dcc_telnet_pass(int idx, int atr)
    * will simply ignore the request and everything will go on as usual.
    */
     if (!dcc[idx].ssl) {
+      /* find number in socklist */
+      int i = findsock(dcc[idx].sock);
+      struct threaddata *td = threaddata();
+      /* mark socket to read next incoming at reduced len */
+      td->socklist[i].flags |= SOCK_SENTTLS;
       dprintf(idx, "starttls\n");
       putlog(LOG_BOTS, "*", "Sent STARTTLS to %s...", dcc[idx].nick);
     }

--- a/src/eggdrop.h
+++ b/src/eggdrop.h
@@ -652,6 +652,10 @@ typedef struct {
 #define SOCK_VIRTUAL    0x0200  /* not-connected socket (dont read it!) */
 #define SOCK_BUFFER     0x0400  /* buffer data; don't notify dcc funcs  */
 #define SOCK_TCL        0x0800  /* tcl socket, don't do anything on it  */
+#ifdef TLS
+#  define SOCK_SENTTLS  0x1000  /* Socket that awaits a starttls in the
+                                 * next read                            */
+#endif
 
 /* Flags to sock_has_data
  */

--- a/src/net.c
+++ b/src/net.c
@@ -919,6 +919,12 @@ int sockread(char *s, int *len, sock_list *slist, int slistmax, int tclonly)
                      ERR_error_string(ERR_get_error(), 0), err);
             x = -1;
           }
+        } else if (slist[i].flags & SOCK_SENTTLS) {
+          /* We are awaiting a reply on our "starttls", only read
+           * strlen("starttls -\n") bytes so we don't accidently
+           * read the Client Hello from the ssl handshake */
+          x = read(slist[i].sock, s, strlen("starttls -\n"));
+          slist[i].flags &= ~SOCK_SENTTLS;
         } else
           x = read(slist[i].sock, s, grab);
       }

--- a/src/tls.c
+++ b/src/tls.c
@@ -780,7 +780,10 @@ int ssl_handshake(int sock, int flags, int verify, int loglevel, char *host,
   SSL_set_mode(td->socklist[i].ssl, SSL_MODE_ENABLE_PARTIAL_WRITE |
                SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER);
   if (data->flags & TLS_CONNECT) {
+    struct timespec req = { 0, 1000000L };
     SSL_set_verify(td->socklist[i].ssl, SSL_VERIFY_PEER, ssl_verify);
+    /* Introduce 1ms lag so an unpatched hub has time to setup the ssl handshake */
+    nanosleep(&req, NULL);
     ret = SSL_connect(td->socklist[i].ssl);
     if (!ret)
       debug0("TLS: connect handshake failed.");

